logstash-core/lib/logstash/plugins/builtin/pipeline/input.rb

```diff
@@ -55,16 +56,22 @@ def running?
   # To understand why this value is useful see Internal.send_to
   # Note, this takes a java Stream, not a ruby array
   def internalReceive(events)
-    return false if !@running.get()
+    return ReceiveResponse.closing() if !@running.get()
 
     # TODO This should probably push a batch at some point in the future when doing so
     # buys us some efficiency
-    events.forEach do |event|
-      decorate(event)
-      @queue << event
+    begin
+      stream_position = 0
+      events.forEach do |event|
+        decorate(event)
+        @queue << event
+        stream_position = stream_position + 1
+      end
+      ReceiveResponse.completed()
+    rescue => e
+      # maybe an IOException in enqueueing
+      ReceiveResponse.failed_at(stream_position, e)
```

Comment:
I was thinking we should narrow down the type rescued if we know we can safely retry only some Java exceptions, but this gets a bit tricky:

- memory queue might get interrupted thus raising a (retryable) `java.lang.InterruptedException`
- the acked queue [maps](https://github.com/elastic/logstash/blob/v8.1.3/logstash-core/src/main/java/org/logstash/ackedqueue/ext/JRubyAckedQueueExt.java#L130=) `java.io.IOException` into an `IOError`

which would lead us to having smt like `rescue java.lang.Exception, IOError => e`

if you've been already considering this and think it's best to just re-try (in which case I am fine with the code as is) ?

in either case we might want some detailed logging at debug level in order to ease if the retry loop runs into a bug, now we only print the `lastResponse.getCauseMessage()`.

Reply:
Agree with you @kares . 
Catching all subclasses of `java.lang.Exception` is it similar to `catch StandardError => e` ? 
So at that point if the only admitted error condition that imply a retry are:
- memory queue interrupted when inserting
- and IOError when wrinting to PQ

shouldn't be
```ruby
rescue java.lang.InterruptedException, IOError
```

and not `java.lang.Exception` ?

Reply:
`rescue java.lang.InterruptedException, IOError` is probably better, haven't looked deep enough and I thought you wanted to catch a bit more (as a precaution) -> on the other hand catching ruby errors (except the known `IOError`) seemed like too much given most of it can only come from the `decorate` invocation.
thus ended up with `java.lang.Exception` :mechanic:

Reply:
Change applied :+1: